File: python/372.py
"""
Project Euler Problem 372: Pencils of Rays

R(M, N) = number of lattice points (x, y) with M < x <= N, M < y <= N
where floor(y^2 / x^2) is odd.

R(0, 100) = 3019, R(100, 10000) = 29750422.
Find R(2*10^6, 10^9).
"""

from math import isqrt


def exact_floor_n_alpha(n, d, P, Q):
    """Compute floor(n * (sqrt(d) + P) / Q) exactly.

    Represents alpha = (sqrt(d) + P) / Q as a quadratic irrational.
    """
    if n == 0:
        return 0
    S = isqrt(n * n * d)
    total_int = n * P + S
    q_div, r = divmod(total_int, Q)
    need = Q - r
    if need <= 0:
        return q_div + 1
    rhs = S + need
    if n * n * d >= rhs * rhs:
        return q_div + 1
    return q_div


def sum_floor(n, d):
    """Compute sum_{x=1}^n floor(x * sqrt(d)) using continued fraction recursion.

    For perfect square d = s^2, returns s * n * (n+1) / 2.
    For non-perfect-square d, uses the lattice point counting recursion:
      SF(n, alpha) = n*m - a*m*(m+1)/2 - SF(m, beta)
    where alpha = sqrt(d) - isqrt(d), m = floor(n*alpha), a = floor(1/alpha),
    and beta = 1/alpha - a.
    """
    if n <= 0:
        return 0
    s = isqrt(d)
    if s * s == d:
        return s * n * (n + 1) // 2

    base = s * n * (n + 1) // 2

    P_cur = -s
    Q_cur = 1
    n_cur = n
    sign = 1
    result = base

    while n_cur > 0:
        m = exact_floor_n_alpha(n_cur, d, P_cur, Q_cur)
        if m == 0:
            break

        denom = d - P_cur * P_cur
        Q_new = denom // Q_cur
        P_inv = -P_cur

        a = exact_floor_n_alpha(1, d, P_inv, Q_new)

        result += sign * (n_cur * m - a * m * (m + 1) // 2)

        P_cur = P_inv - a * Q_new
        Q_cur = Q_new
        n_cur = m
        sign = -sign

    return result


def floor_div_sqrt(num, den):
    """Compute floor(num / sqrt(den)) = largest integer x with x^2 * den <= num^2."""
    if den == 0 or num <= 0:
        return 0
    # x^2 * den <= num^2
    val = num * num // den
    z = isqrt(val)
    # Adjust: we need z^2 * den <= num^2
    while (z + 1) * (z + 1) * den <= num * num:
        z += 1
    while z * z * den > num * num:
        z -= 1
    return z


def sum_upper(ll, rr, N, d):
    """Compute sum of min(N, upper(x, d)) for x in [ll, rr].

    upper(x, d) = floor(sqrt(d*x^2 - 1)):
    - If d = s^2: upper(x,d) = s*x - 1
    - If d is not a perfect square: upper(x,d) = floor(x*sqrt(d)) = isqrt(d*x^2)
    """
    if ll > rr:
        return 0

    s = isqrt(d)
    if s * s == d:
        # upper(x,d) = s*x - 1
        # Split at x where s*x - 1 >= N, i.e., s*x >= N+1, i.e., x >= ceil((N+1)/s)
        x_split = (N + s) // s  # ceil((N+1)/s)
        # For x < x_split: min(N, s*x-1) = s*x-1
        # For x >= x_split: min(N, s*x-1) = N
        left_end = min(rr, x_split - 1)
        res = 0
        if ll <= left_end:
            # sum of (s*x - 1) for x = ll to left_end
            cnt = left_end - ll + 1
            sum_x = cnt * (ll + left_end) // 2
            res += s * sum_x - cnt
        right_start = max(ll, x_split)
        if right_start <= rr:
            cnt = rr - right_start + 1
            res += N * cnt
        return res
    else:
        # upper(x,d) = floor(x*sqrt(d))
        # Split at x where floor(x*sqrt(d)) >= N, i.e., x >= ceil(N/sqrt(d))
        # ceil(N/sqrt(d)) = floor_div_sqrt(N, d) + 1 if N/sqrt(d) is not integer
        # Since d is not perfect square, N/sqrt(d) is never integer for integer N.
        x_split = floor_div_sqrt(N, d) + 1  # smallest x with floor(x*sqrt(d)) >= N
        # Actually: floor(x*sqrt(d)) >= N iff x*sqrt(d) >= N iff x >= N/sqrt(d)
        # So x_split = ceil(N/sqrt(d)) = floor_div_sqrt(N, d) + 1

        # For x < x_split: min(N, floor(x*sqrt(d))) = floor(x*sqrt(d))
        # For x >= x_split: min(N, floor(x*sqrt(d))) = N
        left_end = min(rr, x_split - 1)
        res = 0
        if ll <= left_end:
            res += sum_floor(left_end, d) - sum_floor(ll - 1, d)
        right_start = max(ll, x_split)
        if right_start <= rr:
            cnt = rr - right_start + 1
            res += N * cnt
        return res


def sum_lower_max(ll, rr, L, d):
    """Compute sum of max(L, lower(x, d)) for x in [ll, rr].

    lower(x, d) = ceil(x*sqrt(d)):
    - If d = s^2: lower(x,d) = s*x
    - If d is not a perfect square: lower(x,d) = floor(x*sqrt(d)) + 1
    """
    if ll > rr:
        return 0

    s = isqrt(d)
    if s * s == d:
        # lower(x,d) = s*x
        # max(L, s*x). Split at x where s*x >= L, i.e., x >= ceil(L/s)
        x_split = (L + s - 1) // s  # ceil(L/s)
        res = 0
        # For x < x_split: s*x < L, so max(L, s*x) = L
        left_end = min(rr, x_split - 1)
        if ll <= left_end:
            cnt = left_end - ll + 1
            res += L * cnt
        # For x >= x_split: s*x >= L, so max(L, s*x) = s*x
        right_start = max(ll, x_split)
        if right_start <= rr:
            cnt = rr - right_start + 1
            sum_x = cnt * (right_start + rr) // 2
            res += s * sum_x
        return res
    else:
        # lower(x,d) = floor(x*sqrt(d)) + 1
        # max(L, floor(x*sqrt(d)) + 1)
        # Split at x where floor(x*sqrt(d)) + 1 >= L, i.e., floor(x*sqrt(d)) >= L-1
        # floor(x*sqrt(d)) >= L-1 iff x*sqrt(d) >= L-1 (since L-1 is integer and x*sqrt(d) irrational)
        # iff x >= (L-1)/sqrt(d), x >= ceil((L-1)/sqrt(d)) = floor_div_sqrt(L-1, d) + 1
        if L <= 1:
            x_split = 1  # lower is always >= 1 >= L
        else:
            x_split = floor_div_sqrt(L - 1, d) + 1

        res = 0
        # For x < x_split: floor(x*sqrt(d)) + 1 < L, so max = L
        left_end = min(rr, x_split - 1)
        if ll <= left_end:
            cnt = left_end - ll + 1
            res += L * cnt
        # For x >= x_split: floor(x*sqrt(d)) + 1 >= L, so max = floor(x*sqrt(d)) + 1
        right_start = max(ll, x_split)
        if right_start <= rr:
            cnt = rr - right_start + 1
            sf = sum_floor(rr, d) - sum_floor(right_start - 1, d)
            res += sf + cnt  # sum of (floor(x*sqrt(d)) + 1) = sf + count
        return res


def compute_R(M, N):
    L = M + 1
    if L > N:
        return 0
    total = 0
    max_ratio = N // L + 1
    max_k = 2 * max_ratio * max_ratio
    k = 1
    while k <= max_k:
        # x range: x >= ceil(L/sqrt(k+1)), x <= floor(N/sqrt(k))
        # Also x in [L, N].
        x_lo = floor_div_sqrt(L - 1, k + 1) + 1  # ceil(L/sqrt(k+1)): smallest x with x*sqrt(k+1) >= L
        # Actually: we need lower(x, k) <= N, which means ceil(x*sqrt(k)) <= N.
        # And upper(x, k+1) >= L, which means floor(sqrt((k+1)*x^2-1)) >= L.
        # For x range: x >= ceil(L/sqrt(k+1)) and x <= floor(N/sqrt(k)).
        # But also x >= L and x <= N.
        x_start = max(L, x_lo)
        x_end = min(N, floor_div_sqrt(N, k))

        if x_start > x_end:
            k += 2
            continue

        su = sum_upper(x_start, x_end, N, k + 1)
        sl = sum_lower_max(x_start, x_end, L, k)
        num_x = x_end - x_start + 1
        contrib = su - sl + num_x
        if contrib > 0:
            total += contrib
        k += 2
    return total


if __name__ == "__main__":
    print(compute_R(2000000, 1000000000))

File: c/372.c
/*
 * Project Euler Problem 372: Pencils of Rays
 *
 * R(M, N) = number of lattice points (x, y) with M < x <= N, M < y <= N
 * where floor(y^2 / x^2) is odd.
 *
 * Find R(2*10^6, 10^9).
 *
 * Algorithm: For each odd k, count lattice points (x,y) with floor(y^2/x^2) = k,
 * meaning k*x^2 <= y^2 < (k+1)*x^2, i.e., x*sqrt(k) <= y < x*sqrt(k+1).
 * Sum contributions using floor-sum-of-sqrt via continued fraction recursion.
 */

#include <stdio.h>
#include <math.h>

typedef long long ll;
typedef __int128 lll;

static ll isqrt_ll(ll n) {
    if (n <= 0) return 0;
    ll x = (ll)sqrt((double)n);
    while (x > 0 && x * x > n) x--;
    while ((x + 1) * (x + 1) <= n) x++;
    return x;
}

/* Compute floor(n * (sqrt(d) + P) / Q) exactly */
static ll exact_floor_n_alpha(ll n, ll d, ll P, ll Q) {
    if (n == 0) return 0;
    lll nd = (lll)n * n * d;
    ll S = isqrt_ll((ll)((double)n * sqrt((double)d)));
    /* Refine S so that S^2 <= n^2*d < (S+1)^2 */
    while ((lll)S * S > nd) S--;
    while ((lll)(S + 1) * (S + 1) <= nd) S++;

    lll total_int = (lll)n * P + S;
    ll q_div = (ll)(total_int / Q);
    lll r = total_int - (lll)q_div * Q;
    lll need = Q - r;
    if (need <= 0) return q_div + 1;
    lll rhs = S + need;
    if (nd >= rhs * rhs) return q_div + 1;
    return q_div;
}

/* Compute sum_{x=1}^n floor(x * sqrt(d)) using continued fraction recursion */
static ll sum_floor(ll n, ll d) {
    if (n <= 0) return 0;
    ll s = isqrt_ll(d);
    if (s * s == d) return s * n * (n + 1) / 2;

    ll base = s * n * (n + 1) / 2;
    ll P_cur = -s, Q_cur = 1;
    ll n_cur = n;
    int sign = 1;
    ll result = base;

    while (n_cur > 0) {
        ll m = exact_floor_n_alpha(n_cur, d, P_cur, Q_cur);
        if (m == 0) break;

        lll denom = (lll)d - (lll)P_cur * P_cur;
        ll Q_new = (ll)(denom / Q_cur);
        ll P_inv = -P_cur;

        ll a = exact_floor_n_alpha(1, d, P_inv, Q_new);

        if (sign > 0)
            result += n_cur * m - a * m * (m + 1) / 2;
        else
            result -= n_cur * m - a * m * (m + 1) / 2;

        P_cur = P_inv - a * Q_new;
        Q_cur = Q_new;
        n_cur = m;
        sign = -sign;
    }

    return result;
}

static ll floor_div_sqrt(ll num, ll den) {
    if (den == 0 || num <= 0) return 0;
    lll num2 = (lll)num * num;
    lll z2_den = num2 / den;
    ll z = isqrt_ll((ll)((double)num / sqrt((double)den)));
    while ((lll)(z + 1) * (z + 1) * den <= num2) z++;
    while ((lll)z * z * den > num2) z--;
    return z;
}

static ll sum_upper(ll ll_v, ll rr, ll N_val, ll d) {
    if (ll_v > rr) return 0;
    ll s = isqrt_ll(d);
    if (s * s == d) {
        ll x_split = (N_val + s) / s;
        ll res = 0;
        ll left_end = rr < x_split - 1 ? rr : x_split - 1;
        if (ll_v <= left_end) {
            ll cnt = left_end - ll_v + 1;
            ll sum_x = cnt * (ll_v + left_end) / 2;
            res += s * sum_x - cnt;
        }
        ll right_start = ll_v > x_split ? ll_v : x_split;
        if (right_start <= rr) {
            ll cnt = rr - right_start + 1;
            res += N_val * cnt;
        }
        return res;
    } else {
        ll x_split = floor_div_sqrt(N_val, d) + 1;
        ll left_end = rr < x_split - 1 ? rr : x_split - 1;
        ll res = 0;
        if (ll_v <= left_end) {
            res += sum_floor(left_end, d) - sum_floor(ll_v - 1, d);
        }
        ll right_start = ll_v > x_split ? ll_v : x_split;
        if (right_start <= rr) {
            ll cnt = rr - right_start + 1;
            res += N_val * cnt;
        }
        return res;
    }
}

static ll sum_lower_max(ll ll_v, ll rr, ll L_val, ll d) {
    if (ll_v > rr) return 0;
    ll s = isqrt_ll(d);
    if (s * s == d) {
        ll x_split = (L_val + s - 1) / s;
        ll res = 0;
        ll left_end = rr < x_split - 1 ? rr : x_split - 1;
        if (ll_v <= left_end) {
            ll cnt = left_end - ll_v + 1;
            res += L_val * cnt;
        }
        ll right_start = ll_v > x_split ? ll_v : x_split;
        if (right_start <= rr) {
            ll cnt = rr - right_start + 1;
            ll sum_x = cnt * (right_start + rr) / 2;
            res += s * sum_x;
        }
        return res;
    } else {
        ll x_split;
        if (L_val <= 1) {
            x_split = 1;
        } else {
            x_split = floor_div_sqrt(L_val - 1, d) + 1;
        }
        ll res = 0;
        ll left_end = rr < x_split - 1 ? rr : x_split - 1;
        if (ll_v <= left_end) {
            ll cnt = left_end - ll_v + 1;
            res += L_val * cnt;
        }
        ll right_start = ll_v > x_split ? ll_v : x_split;
        if (right_start <= rr) {
            ll cnt = rr - right_start + 1;
            ll sf = sum_floor(rr, d) - sum_floor(right_start - 1, d);
            res += sf + cnt;
        }
        return res;
    }
}

static ll compute_R(ll M, ll N_val) {
    ll L_val = M + 1;
    if (L_val > N_val) return 0;
    ll total = 0;
    ll max_ratio = N_val / L_val + 1;
    ll max_k = 2 * max_ratio * max_ratio;
    ll k = 1;
    while (k <= max_k) {
        ll x_lo = floor_div_sqrt(L_val - 1, k + 1) + 1;
        ll x_start = L_val > x_lo ? L_val : x_lo;
        ll x_end_bound = floor_div_sqrt(N_val, k);
        ll x_end = N_val < x_end_bound ? N_val : x_end_bound;

        if (x_start > x_end) {
            k += 2;
            continue;
        }

        ll su = sum_upper(x_start, x_end, N_val, k + 1);
        ll sl = sum_lower_max(x_start, x_end, L_val, k);
        ll num_x = x_end - x_start + 1;
        ll contrib = su - sl + num_x;
        if (contrib > 0) total += contrib;
        k += 2;
    }
    return total;
}

int main(void) {
    printf("%lld\n", compute_R(2000000, 1000000000));
    return 0;
}
